### PR TITLE
Incremental processing for Kotlin sources

### PR DESCRIPTION
**Pull request by ting-yuan**
For now, changes in other modules, external libraries and Java sources will trigger a full rebuild.

It is disabled by default and can be switched on by the Gradle property `ksp.incremental=true`.

### CODE REVIEW

File: compiler-plugin/src/main/kotlin/com/google/devtools/ksp/Incremental.kt [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+        logSourceToOutputs()
+
+        sourceToOutputsMap.flush(false)
+        sourceToOutputsMap.close()
+    }
+
+    // TODO: Recover if processing failed.
+    fun updateOutputs(outputs: Set<File>, cleanOutputs: Collection<File>) {
+        val outRoot = options.kspOutputDir
+        val bakRoot = File(options.cachesDir, "backups")
```
Comment (neetopia) [inline compiler-plugin/src/main/kotlin/com/google/devtools/ksp/Incremental.kt:356]: Backups seems to be unused.

Comment (ting-yuan) [reply compiler-plugin/src/main/kotlin/com/google/devtools/ksp/Incremental.kt:360]: Added comments in code.

File: compiler-plugin/src/main/kotlin/com/google/devtools/ksp/Incremental.kt [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+        val mutableSourceToOutputs = mutableMapOf<File, MutableSet<File>>().apply {
+            sourceToOutputs.forEach {
+                set(it.key, it.value.toMutableSet())
+            }
+        }
+
+        // Associate unassociated outputs to ALL FILES.
+        val associated = sourceToOutputs.values.flatten().toSet()
+        val unassociated = outputs.filterNot { it in associated }
+        mutableSourceToOutputs.getOrPut(anyChangesWildcard) { mutableSetOf() }.addAll(unassociated)
```
[the remark this answers is not in the document]
Comment (ting-yuan) [reply compiler-plugin/src/main/kotlin/com/google/devtools/ksp/Incremental.kt:341]: There are several choices of default association.

1. Associate with everything. This is probably safest but also a performance killer for non-cooperating processors.

2. Associate with nothing. They don't affect the dirty set. This somehow forces processors to provide associations. Valid use cases without association include log / debug files.

I much prefer 2. but 1. seems to be easier to start with. I'll pitch the question in discussion channels. Before making a decision, this implementation is easiest to adapt to either choice.

**Review by neetopia — APPROVED** [review]
Feel free to clean commit history and push.

**Pull request merged**